### PR TITLE
[ENH] Add to_markdown method

### LINKED ISSUE #11052: DataFrame could have a to_markdown method.
Similar to `to_latex` and `to_html`.

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #11052 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -72,6 +72,7 @@ dependencies:
   - matplotlib>=2.2.2  # pandas.plotting, Series.plot, DataFrame.plot
   - numexpr>=2.6.8
   - scipy>=1.1
+  - tabulate
```
Comment (jbrockmendel) [inline environment.yml]: should this go in the "optional for io" block below?  should any particular minimum version be required?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1964,6 +1965,29 @@ def to_feather(self, path):
...
+        --------
+        >>> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
+        >>> print(df.to_markdown())
+        |    |   col1 |   col2 |
+        |---:|-------:|-------:|
+        |  0 |      1 |      3 |
+        |  1 |      2 |      4 |
+        """
+        tabulate = import_optional_dependency("tabulate")
+        return self.pipe(tabulate.tabulate, headers="keys", tablefmt="pipe")
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: id have to go look up how `pipe` works.  is there a lower-tech way of doing this?

File: pandas/tests/io/formats/test_to_markdown.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import pandas.util._test_decorators as td
+
+import pandas as pd
+
+
+@td.skip_if_no_tabulate
```
Comment (jbrockmendel) [inline pandas/tests/io/formats/test_to_markdown.py]: this can probably use a pytestmark at the module level

**Comment by jbrockmendel** [comment]
should Series support this too?  does `tabulate` have something that could be used for a `read_markdown` to round-trip?

File: pandas/core/frame.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/core/frame.py]: `tabulate.tabulate(self, headers="keys", tablefmt="pipe")` I suspect.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -1964,6 +1965,29 @@ def to_feather(self, path):
 
         to_feather(self, path)
 
+    def to_markdown(self):
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Should `**kwargs` be passed through to tabulate?

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -207,6 +207,7 @@ Other enhancements
 - The ``partition_cols`` argument in :meth:`DataFrame.to_parquet` now accepts a string (:issue:`27117`)
 - :func:`to_parquet` now appropriately handles the ``schema`` argument for user defined schemas in the pyarrow engine. (:issue: `30270`)
 - DataFrame constructor preserve `ExtensionArray` dtype with `ExtensionArray` (:issue:`11363`)
+- :meth:`DataFrame.to_markdown` added (:issue:`11052`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: Add this to `source/reference/frame.rst`.

**Review by TomAugspurger — COMMENTED** [review]
You'll need to add tabulate to at least one of the `ci/deps/` files

File: pandas/core/series.py [hunk 6]
```diff
@@ -1434,6 +1434,28 @@ def to_string(
...
+
+        Examples
+        --------
+        >>> s = pd.Series([1, 2, 3, 4])
+        >>> print(s.to_markdown())
+        |    |   col1 |   col2 |
+        |---:|-------:|-------:|
+        |  0 |      1 |      3 |
+        |  1 |      2 |      4 |
+        """
```
Comment (MarcoGorelli) [inline pandas/core/series.py]: This isn't right, will update it now

File: pandas/tests/io/formats/test_to_markdown.py (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/tests/io/formats/test_to_markdown.py]: Does that mean to make a class and then apply the decorator to the whole class? If so, that's what I've done

File: pandas/core/frame.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py]: Sure, done

File: ci/deps/travis-38.yaml [hunk 7]
```diff
@@ -17,3 +17,4 @@ dependencies:
   - nomkl
   - pytz
   - pip
+  - tabulate==0.8.6
```
Comment (MarcoGorelli) [inline ci/deps/travis-38.yaml]: I'll check how far back we can go without breaking the tests

**Comment by MarcoGorelli** [comment]
> should Series support this too? does `tabulate` have something that could be used for a `read_markdown` to round-trip?

Sure, I've added it to Series.

I haven't found anything in the `tabulate` source code that can be easily used for `read_markdown`, but it would be a nice enhancement to work on

**Comment by darkdragon-001** [comment]
> I haven't found anything in the `tabulate` source code that can be easily used for `read_markdown`, but it would be a nice enhancement to work on.

Maybe CSV code with `|` separator could work. One just has to ignore the header separation line. Maybe empty columns will be added because of the outer border separators and should be removed in post-processing automatically.

File: pandas/core/frame.py [hunk 8]
```diff
@@ -1964,6 +1965,33 @@ def to_feather(self, path):
 
         to_feather(self, path)
 
+    def to_markdown(self, **kwargs):
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I think my preferred signature here is something like

```
def to_markdown(self, buf=None, mode=None, **kwargs):
    ...
```

In the future, we may have an `engine` keyword for other formatting libraries like tabulate, so we should probably avoid picking too many keywords (like tblfmt, headers, etc.).  

`buf` will have the same semantics as `to_string()` / `to_csv()`. We print to stdout by default but you can also pass your own buffer or filename as a string.

I suspect people will want an option for `index=True/False`. Not sure if we'll want to support that yet.

File: pandas/core/frame.py [hunk 9]
```diff
@@ -1964,6 +1965,33 @@ def to_feather(self, path):
...
+        --------
+        >>> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
+        >>> print(df.to_markdown())
+        |    |   col1 |   col2 |
+        |---:|-------:|-------:|
+        |  0 |      1 |      3 |
+        |  1 |      2 |      4 |
+        """
+        tabulate = import_optional_dependency("tabulate")
+        if "headers" not in kwargs:
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: This and the next line can be

```suggestion
        kwargs.setdefault("headers", "keys")
```

likewise for tablefmt.

File: pandas/tests/io/formats/test_to_markdown.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/tests/io/formats/test_to_markdown.py]: It'd be `pytestmark = pytest.importorskip("tabulate")` But this is fine too.

File: pandas/core/frame.py (hunk 8 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py]: Sure, I've tried updating this as requested.

I haven't added a `pandas/io/formats/markdown.py` file - would that be necessary?

Comment (TomAugspurger) [reply pandas/core/frame.py]: Probably not necessary yet.

File: ci/deps/travis-38.yaml [hunk 10]
```diff
@@ -17,3 +17,4 @@ dependencies:
   - nomkl
   - pytz
   - pip
+  - tabulate==0.8.0
```
Comment (TomAugspurger) [inline ci/deps/travis-38.yaml]: We should job pinned to the oldest version and one that's unpinned.

File: pandas/core/frame.py [hunk 11]
```diff
@@ -1964,6 +1966,40 @@ def to_feather(self, path):
...
+    def to_markdown(
+        self,
+        buf: Optional[FilePathOrBuffer[str]] = None,
+        mode: Optional[str] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Print a DataFrame in markdown-friendly format.
+
+        .. versionadded:: 1.0
```
Comment (TomAugspurger) [inline pandas/core/frame.py]:
```suggestion
        .. versionadded:: 1.0.0
```

File: pandas/core/frame.py [hunk 12]
```diff
@@ -1964,6 +1966,40 @@ def to_feather(self, path):
...
+
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Add a `Parameters` section here.

File: pandas/core/frame.py [hunk 13]
```diff
@@ -1964,6 +1966,40 @@ def to_feather(self, path):
...
+        Examples
+        --------
+        >>> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
+        >>> print(df.to_markdown())
+        |    |   col1 |   col2 |
+        |---:|-------:|-------:|
+        |  0 |      1 |      3 |
+        |  1 |      2 |      4 |
+        """
+        if buf is None:
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Does this work? I wouldn't think `get_filepath_or_buffer` would work with sys.stdout with different modes.

I'd look at how CSVFormatter does this. It creates an `io.StringIO` object and passes that to `get_filepath_or_buffer`.

And when it's `None`, then we return `buf.getvalue()`

File: pandas/core/series.py [hunk 14]
```diff
@@ -1434,6 +1435,35 @@ def to_string(
                 with open(buf, "w") as f:
                     f.write(result)
 
+    def to_markdown(
+        self,
+        buf: Optional[FilePathOrBuffer[str]] = None,
+        mode: Optional[str] = None,
+        **kwargs,
+    ) -> str:
+        """
```
Comment (TomAugspurger) [inline pandas/core/series.py]: Maybe worth doing a shared doc.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add to install.rst and pandas/compat/_optional.py

File: pandas/tests/io/formats/test_to_markdown.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    assert result == (
+        "|    |   foo |\n|---:|------:|\n|  0 |     1 "
+        "|\n|  1 |     2 |\n|  2 |     3 |"
+    )
+
+
+def test_no_buf(capsys):
+    df = pd.DataFrame([1, 2, 3])
+    df.to_markdown()
+    out, _ = capsys.readouterr()
```
Comment (TomAugspurger) [inline pandas/tests/io/formats/test_to_markdown.py]: I'm not sure that this is what we want (apologies if misled you earlier).

Rather, we'll want to return a string like `to_csv()` does.

File: pandas/core/frame.py (hunk 13 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py]: Ah OK, thanks.

But is it necessary to return `buf.getvalue()`, or would something like this be enough?
```
        result = tabulate.tabulate(self, **kwargs)
        if buf is None:
            return result
        buf, _, _, _ = get_filepath_or_buffer(buf, mode=mode)
        buf.writelines(result)
```

Comment (TomAugspurger) [reply pandas/core/frame.py]: That's probably fine too.

File: pandas/core/frame.py (hunk 12 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py]: Sure, done. Wasn't sure what to write for `mode`, as the docstring for `pandas/io/common.py::get_filepath_or_buffer` just says
```
mode : str, optional
```
and, as far as I can tell, the parameter is only used in the following conditions:
```
if is_s3_url(filepath_or_buffer):
if is_gcs_url(filepath_or_buffer):
```

**Review by datapythonista — COMMENTED** [review]
Thanks @MarcoGorelli, this is quite useful.

But personally I think this is a great opportunity to implement this as a third party module, and start to work on making it easier to extend pandas with I/O adapters like this (see #26804)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, can you merge master & add a mention in install.rst (and link to the tabulate docs). ping on green.

**Comment by MarcoGorelli** [comment]
@jreback I've mentioned tabulate and added a link to its docs in `install.rst`, then rebased

@datapythonista Thanks! I've read through the thread you linked to. There seems to still be some disagreement among the core team about the issue though, so I'm not sure what to do and am in no position to pick sides

**Comment by jreback** [comment]
> Thanks @MarcoGorelli, this is quite useful.
> 
> But personally I think this is a great opportunity to implement this as a third party module, and start to work on making it easier to extend pandas with I/O adapters like this (see #26804)

@datapythonista that's orthogonal, and to be honest *this* IS almost entirely a third-party module.

It is *much* *much* easier to simply refactor later than to re-factor while adding something new.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli very nice!

**Pull request merged**